compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
         OptimizationFinder { body, tcx, optimizations: OptimizationList::default() }
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
+        // _2 = &_1;
+        // ...
+        // _5 = (*_2);
+        //
+        // which we can replace the last statement with `_5 = _1;` to avoid the load of _2
+        if let Rvalue::Use(op) = rvalue {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
+            let lower_index = stmt_index.saturating_sub(6);
```

Comment:
I believe the only reason my example does not get misoptimized is this magic number. That is too fragile for my liking and this fact is also not documented.